pkg/wslistener/wslistener.go

```diff
@@ -0,0 +1,97 @@
+package wslistener
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/beekeeper/pkg/bee"
+	"github.com/ethersphere/beekeeper/pkg/logging"
+	"github.com/gorilla/websocket"
+)
+
+// ListenWebSocket listens for messages on a websocket connection.
+func ListenWebSocket(ctx context.Context, client *bee.Client, endpoint string, logger logging.Logger) (<-chan string, func(), error) {
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 45 * time.Second,
+	}
+
+	ws, _, err := dialer.DialContext(ctx, fmt.Sprintf("ws://%s%s", client.Host(), endpoint), http.Header{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ch := make(chan string)
+	readCh := make(chan []byte)
+	errCh := make(chan error)
+	done := make(chan struct{})
+
+	go func() {
+		defer close(readCh)
+		defer close(errCh)
+
+		for {
+			// ping should be received every 60s (defined on bee)
+			err := ws.SetReadDeadline(time.Now().Add(90 * time.Second))
+			if err != nil {
+				errCh <- err
+				return
+			}
+			_, data, err := ws.ReadMessage()
+			if err != nil {
+				errCh <- err
+				return
+			}
+			select {
+			case readCh <- data:
+			case <-done:
+				return
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	go func() {
+		defer close(ch)
+
+		for {
+			select {
+			case data := <-readCh:
+				logger.WithField("node", client.Name()).Infof("websocket received message: %s", string(data))
+				select {
+				case ch <- string(data):
+				case <-done:
+					return
+				}
+			case err := <-errCh:
+				logger.Infof("websocket error: %v", err)
+				return
+			case <-ctx.Done():
+				logger.Info("context canceled, closing websocket")
+				return
+			case <-done:
+				return
+			}
+		}
+	}()
+
+	var once sync.Once
+	closer := func() {
+		once.Do(func() {
+			deadline := time.Now().Add(5 * time.Second)
+			msg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+			if err := ws.WriteControl(websocket.CloseMessage, msg, deadline); err != nil {
+				logger.Infof("failed to send close message: %v", err)
+			}
+
+			close(done)
+			ws.Close()
```

Comment:
Maybe it is better first to close the connection and than to close channel, also, `ws.Close()` returns an error, so why not log it if it happens.

Reply:
closing the socket first can result in a `read on closed connection`.
So we close the channel first to ensure that the goroutine returns. Then close the underlying connection